fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/leaderboard/TimeWindowLeaderboardIndexMaintainer.java

```diff
@@ -167,29 +199,45 @@ protected void saveDirectory(TimeWindowLeaderboardDirectory directory) {
             });
         }
         // Add leaderboard's key to the front and take it off of the results.
-        return RecordCursor.fromFuture(getExecutor(), scoreRangeFuture)
-                .flatMapPipelined(scoreRange -> {
-                    if (scoreRange == null) {
-                        return RecordCursor.empty(getExecutor());
-                    } else if (scanType == IndexScanType.BY_VALUE && leaderboardFuture.join().isHighScoreFirst()) {
+        return RecordCursor.fromFuture(getExecutor(), scoreRangeFuture).flatMapPipelined(scoreRange -> {
+            if (scoreRange == null) {
+                return RecordCursor.empty(getExecutor());
+            }
+            final TimeWindowLeaderboard leaderboard = leaderboardFuture.join(); // Already waited in scoreRangeFuture.
```

Comment:
I think that only instruments `asyncToSync`.

Reply:
Yeah, `FDBDatabaseFactory.join` is *almost* sufficient, but it will do the wrong thing if one configures async detection to warn or error on an already completed future. I filed an issue about this: #641.